jina/peapods/stream/gateway.py

```diff
@@ -1,20 +1,86 @@
+from ..runtimes.gateway.graph.topology_graph import TopologyGraph
 
-__all__ = ['ZmqGatewayStreamer', 'GrpcGatewayStreamer']
-
-if TYPE_CHECKING:
-    from ..grpc import Grpclet
-    from ...types.request import Request, Response
+__all__ = ['GatewayStreamer']
 
 
 class GatewayStreamer(BaseStreamer):
     """Streamer used at Gateway to stream requests/responses to/from Executors"""
 
+    def __init__(self, args, graph: TopologyGraph, **kwargs):
+        super().__init__(args, **kwargs)
+        self._graph = graph
+
+    def _handle_request(self, request: 'Request') -> 'asyncio.Future':
+        """
+        For zmq & grpc data requests from gateway, for each request in the iterator, we send the `Message`
+        using `iolet.send_message()`.
+
+        For websocket requests from client, for each request in the iterator, we send the request in `bytes`
+        using using `iolet.send_message()`.
+
+        Then add {<request-id>: <an-empty-future>} to the request buffer.
+        This empty future is used to track the `result` of this request during `receive`.
+
+        :param request: current request in the iterator
+        :return: asyncio Future for sending message
+        """
+        graph = copy.deepcopy(self._graph)
+        # important that the gateway needs to have an instance of the graph per request
+        tasks_to_respond = []
+        tasks_to_ignore = []
+        for origin_node in graph.origin_nodes:
+            leaf_tasks = origin_node.get_leaf_tasks(
+                self._connection_pool, self._convert_to_message(request), None
+            )
+            # Every origin node returns a set of tasks that are the ones corresponding to the leafs of each of their subtrees that unwrap all the previous tasks.
+            # It starts like a chain of waiting for tasks from previous nodes
+            tasks_to_respond.extend([task for ret, task in leaf_tasks if ret])
+            tasks_to_ignore.extend([task for ret, task in leaf_tasks if not ret])
+
+        async def _merge_results_at_end_gateway(
+            tasks: List[asyncio.Task],
+        ) -> asyncio.Future:
+            from jina import DocumentArray
+
+            # TODO: Should the order be deterministic by the graph structure, or depending on the response speed?
+            # partial_responses = [await result for result in asyncio.as_completed(*tasks)]
+            partial_responses = await asyncio.gather(*tasks)
+
+            # when merging comes, one task may return None
+            filtered_partial_responses = list(
+                filter(lambda x: x is not None, partial_responses)
+            )
+            if len(filtered_partial_responses) > 1:
+                docs = DocumentArray(
+                    [
+                        d
+                        for r in filtered_partial_responses
+                        for d in getattr(r.request, 'docs')
+                    ]
+                )
+                groundtruths = DocumentArray(
```

Comment:
we didnt auto merge the groundtruths before, did we? (only in cases where we had some executor I think?)

Reply:
yes, which seems a little bit dangerous, We can revisit this later